README.md
# Resolved 🌐

Learn how to build (and test) a [production ready][1] Rack application from
scratch.

![A screenshot of the homepage. It shows the DNS records for thoughtbot.com](./demo.png)

## Topics Covered 📚

Each [commit][2] serves as a mini lesson.

- Rendering dynamic content with `ERB`.
- Using HTTP compression and caching to drastically improve application
  performance.
- Error handling.
- Building a simple logging mechanism.
- HTTP security best practices.
- Configuring a test suite and GitHub Actions.

## Setup 🚀

```
bundle install
```

## Local Development 🧱

```
bin/dev
```

## CI ✅

```
bundle exec rake
```

## Contributing

See the [CONTRIBUTING] document.
Thank you, [contributors]!

  [CONTRIBUTING]: CONTRIBUTING.md
  [contributors]: https://github.com/thoughtbot/resolved/graphs/contributors

## License

Resolved is Copyright (c) thoughtbot, inc.
It is free software, and may be redistributed
under the terms specified in the [LICENSE] file.

  [LICENSE]: LICENSE

## About

![thoughtbot](https://thoughtbot.com/brand_assets/93:44.svg)

Resolved is maintained and funded by thoughtbot, inc.
The names and logos for thoughtbot are trademarks of thoughtbot, inc.

We love open source software!
See [our other projects][community]
or [hire us][hire] to help build your product.

  [community]: https://thoughtbot.com/community?utm_source=github
  [hire]: https://thoughtbot.com/hire-us?utm_source=github

[1]: https://resolved-9623c76e6dbd.herokuapp.com
[2]: https://github.com/thoughtbot/resolved/commits/main


NEWS.md
# News

All notable changes to this project will be documented in this file.

The format is loosely based on [Keep a Changelog] and this project adheres to
[Semantic Versioning].

  [Keep a Changelog]: http://keepachangelog.com/
  [Semantic Versioning]: http://semver.org/

## 2023-08-18

### Added

- Initial release to support blog post.


spec/spec_helper.rb
# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# See https://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
ENV["RACK_ENV"] ||= "test"
require "capybara/rspec"
require_relative "./../app/app"

# https://github.com/teamcapybara/capybara#setup
Capybara.app = App.new

RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

  # This option will default to `:apply_to_host_groups` in RSpec 4 (and will
  # have no way to turn it off -- the option exists only for backwards
  # compatibility in RSpec 3). It causes shared context metadata to be
  # inherited by the metadata hash of host groups and examples, rather than
  # triggering implicit auto-inclusion in groups with matching metadata.
  config.shared_context_metadata_behavior = :apply_to_host_groups

  # The settings below are suggested to provide a good initial experience
  # with RSpec, but feel free to customize to your heart's content.
  #   # This allows you to limit a spec run to individual examples or groups
  #   # you care about by tagging them with `:focus` metadata. When nothing
  #   # is tagged with `:focus`, all examples get run. RSpec also provides
  #   # aliases for `it`, `describe`, and `context` that include `:focus`
  #   # metadata: `fit`, `fdescribe` and `fcontext`, respectively.
  #   config.filter_run_when_matching :focus
  #
  #   # Allows RSpec to persist some state between runs in order to support
  #   # the `--only-failures` and `--next-failure` CLI options. We recommend
  #   # you configure your source control system to ignore this file.
  #   config.example_status_persistence_file_path = "spec/examples.txt"
  #
  #   # Limits the available syntax to the non-monkey patched syntax that is
  #   # recommended. For more details, see:
  #   # https://rspec.info/features/3-12/rspec-core/configuration/zero-monkey-patching-mode/
  #   config.disable_monkey_patching!
  #
  #   # This setting enables warnings. It's recommended, but in some cases may
  #   # be too noisy due to issues in dependencies.
  #   config.warnings = true
  #
  #   # Many RSpec users commonly either run the entire suite or an individual
  #   # file, and it's useful to allow more verbose output when running an
  #   # individual spec file.
  #   if config.files_to_run.one?
  #     # Use the documentation formatter for detailed output,
  #     # unless a formatter has already been configured
  #     # (e.g. via a command-line flag).
  #     config.default_formatter = "doc"
  #   end
  #
  #   # Print the 10 slowest examples and example groups at the
  #   # end of the spec run, to help surface which specs are running
  #   # particularly slow.
  #   config.profile_examples = 10
  #
  #   # Run specs in random order to surface order dependencies. If you find an
  #   # order dependency and want to debug it, you can fix the order by providing
  #   # the seed, which is printed after each run.
  #   #     --seed 1234
  #   config.order = :random
  #
  #   # Seed global randomization in this process using the `--seed` CLI option.
  #   # Setting this allows you to use `--seed` to deterministically reproduce
  #   # test failures related to randomization by passing the same `--seed` value
  #   # as the one that triggered the failure.
  #   Kernel.srand config.seed
end


spec/features/app_spec.rb
require "spec_helper"

feature "App" do
  scenario "Visiting the homepage" do
    visit "/"

    expect(page).to have_content "Resolved"
    expect(page).to have_content "Enter a domain name"
  end

  scenario "Visiting a page that does not exist" do
    visit "/invalid/path"

    expect(page).to have_content "Page not found"
  end

  scenario "Filling out form" do
    resolver = instance_double(Resolv::DNS)
    ns_resource = instance_double(Resolv::DNS::Resource::IN::NS)
    allow(Resolv::DNS).to receive(:new).and_return(resolver)
    allow(resolver).to receive(:getresources).and_return([ns_resource])
    allow(ns_resource).to receive(:name).and_return("expected-server.net")

    visit "/"

    fill_in :url, with: "https://example.com"
    click_button "Submit"

    expect(page).to have_field("url", with: "https://example.com")
    within("ul") do
      expect(page).to have_selector("li", text: "expected-server.net")
    end
  end

  scenario "Handling errors" do
    error = Resolv::ResolvError.new("expected error")
    allow(Resolv::DNS).to receive(:new).and_raise(error)

    visit "/"

    fill_in :url, with: "https://example.com"
    click_button "Submit"

    expect(page).to have_content "expected error"
    expect(page).to have_field("url", with: "https://example.com")

    visit current_path

    expect(page).to_not have_content "expect error"
  end
end


spec/requests/app_spec.rb
require "spec_helper"

RSpec.describe App, type: :request do
  include Rack::Test::Methods

  def app
    Rack::Builder.parse_file("config.ru")
  end

  describe "/" do
    it "returns a 200 status" do
      get "/"

      expect(last_response.status).to eq 200
    end

    it "is compressed" do
      get "/", {}, {"HTTP_ACCEPT_ENCODING" => "gzip"}

      expect(last_response.headers["Content-Encoding"]).to eq "gzip"
    end
  end

  describe "invalid paths" do
    it "returns a 404 status " do
      get "/invalid/path"

      expect(last_response.status).to eq 404
    end

    it "returns an html content type" do
      get "/invalid/path"

      expect(last_response.headers["Content-Type"]).to eq "text/html; charset=utf-8"
    end
  end

  describe "/404.html" do
    it "returns a 200 status" do
      get "/404.html"

      expect(last_response.status).to eq 200
    end

    it "returns an html content type" do
      get "/404.html"

      expect(last_response.headers["Content-Type"]).to eq "text/html; charset=utf-8"
    end
  end

  describe "/?url=" do
    it "returns the correct response" do
      resolver = instance_double(Resolv::DNS)
      ns_resource = instance_double(Resolv::DNS::Resource::IN::NS)
      allow(Resolv::DNS).to receive(:new).and_return(resolver)
      allow(resolver).to receive(:getresources).and_return([ns_resource])
      allow(ns_resource).to receive(:name).and_return("server.net")

      get "/?url=https://example.com"

      expect(last_response.status).to eq 200
      expect(last_response.headers["Content-Type"]).to eq "text/html; charset=utf-8"
    end

    context "when the server throws an error" do
      it "returns a 500 status" do
        allow(Resolv::DNS).to receive(:new).and_raise(StandardError, "some error")

        get "/?url=https://example.com"

        expect(last_response.status).to eq 500
        expect(last_response.headers["Content-Type"]).to eq "text/html; charset=utf-8"
        expect(last_response.body).to match "Something went wrong."
      end
    end

    context "when there are no results" do
      it "returns a 422 status" do
        resolver = instance_double(Resolv::DNS)
        allow(Resolv::DNS).to receive(:new).and_return(resolver)
        allow(resolver).to receive(:getresources).and_return([])

        get "/?url=https://no-name-servers.com"

        expect(last_response.status).to eq 422
        expect(last_response.body).to match "Could not resolve DNS records for no-name-servers.com"
      end
    end
  end

  describe "/styles.css" do
    it "is cached" do
      get "/css/styles.css"

      expect(last_response.headers["Cache-Control"]).to eq "public, max-age=31556952"
    end

    it "is compressed" do
      get "/css/styles.css", {}, {"HTTP_ACCEPT_ENCODING" => "gzip"}

      expect(last_response.headers["Content-Encoding"]).to eq "gzip"
    end
  end

  describe "invalid methods" do
    it "returns a 405 status" do
      post "/"

      expect(last_response.status).to eq 405
    end

    it "returns the correct content type" do
      post "/"

      expect(last_response.headers["Content-Type"]).to eq "text/html; charset=utf-8"
    end

    it "returns the correct response" do
      post "/"

      expect(last_response.body).to match "Method not allowed"
    end
  end

  describe "/405.html" do
    it "returns a 200 status" do
      get "/405.html"

      expect(last_response.status).to eq 200
    end

    it "returns an html content type" do
      get "/405.html"

      expect(last_response.headers["Content-Type"]).to eq "text/html; charset=utf-8"
    end
  end
end


app/app.rb
require "bundler"
require "erb"
require "resolv"
require "uri"

Bundler.require(:default, ENV.fetch("RACK_ENV").to_sym)

class App
  HEADERS = {
    "Content-Type" => "text/html; charset=utf-8",
    "Strict-Transport-Security" => "max-age=63072000; includeSubDomains; preload",
    "X-Content-Type-Options" => "nosniff",
    "X-Frame-Options" => "SAMEORIGIN",
    "Referrer-Policy" => "strict-origin-when-cross-origin"
  }.freeze

  def call(env)
    req = Rack::Request.new(env)
    path = req.path_info

    case path
    when "/"
      url = req.params["url"]

      if url
        result = name_servers_for(url)

        if result.success
          render("home", url:, name_servers: result.payload)
        else
          render("home", url:, announcement: result.error, status_code: 422)
        end
      else
        render("home")
      end
    else
      handle_missing_path
    end
  end

  private

  def render(template, status_code: 200, announcement: nil, headers: {}, **locals)
    @locals = locals
    @announcement = announcement
    @content = render_template(template)
    body = render_template("layout")
    headers = HEADERS.merge(headers)

    [status_code, headers, [body]]
  end

  def render_template(template)
    template = File.read("./app/views/#{template}.html.erb")
    erb = ERB.new(template)
    erb.result(binding)
  end

  def handle_missing_path
    body = File.read("./public/404.html")
    headers = {"Content-Type" => "text/html; charset=utf-8"}

    [404, headers, [body]]
  end

  def name_servers_for(url)
    result = Struct.new(:success, :payload, :error, keyword_init: true)

    begin
      host = URI(url).host
      res = Resolv::DNS.new
      payload = res.getresources(host, Resolv::DNS::Resource::IN::NS)
      raise Resolv::ResolvError, "Could not resolve DNS records for #{host}" if payload.empty?
      sorted_payload = payload.sort_by { |record| record.name.to_s }

      result.new(success: true, payload: sorted_payload)
    rescue Resolv::ResolvError, URI::InvalidURIError => error
      result.new(success: false, error: error.message)
    end
  end
end
